### PR TITLE
[DEV] torch_dist fixes, speed improvements and memory reduction for layerwise distopt

### PR DESCRIPTION
**Pull request by FDecaYed**
# What does this PR do ?

[PR to main](https://github.com/NVIDIA/Megatron-LM/pull/2241)

Adding some optimizations that misses initial PR adding layerwise dist opt:  
- switch to faster all gather
- sort param size and assign to rank in a ping-pong loop for balanced memory usage

Also added 3 minor fixes for layerwise:
- for dist ckpt, correctly set Tensor Factory
- for unwrap accidentally wrapped adam optimizers(FP32) before chained
- handle empty group after sharding when generate sharded_state_dict for ckpting

### Code review

The following process is enforced via the CODEOWNERS file for changes into `megatron/core`. For changes outside of `megatron/core`, it is up to the PR author whether or not to tag the Final Reviewer team.

<details>
<summary>For MRs into `main` branch</summary>

#### (Step 2): Collect the expert reviewers reviews

1. Attach the `Expert Review` label when your PR is ready for review.
2. GitHub auto-assigns expert reviewers based on your changes. They will get notified and pick up your PR soon.

:warning: Only proceed to the next step once all reviewers have approved, merge-conflict are resolved and the CI is passing.  
Final Review might get declined if these requirements are not fulfilled.

#### (Step 3): Final Review

1. Add `Final Review` label
2. GitHub auto-assigns final reviewers based on your changes. They will get notified and pick up your PR soon.

#### (Optional Step 4): Cherry-pick into release branch

If this PR also needs to be merged into `core_r*` release branches, after this PR has been merged, select `Cherry-pick` to open a new PR into the release branch.

</details>

<details>
<summary>For MRs into `dev` branch</summary>
The proposed review process for `dev` branch is under active discussion.

MRs are mergable after one approval by either `eharper@nvidia.com` or `zijiey@nvidia.com`.
</details>

### Merging your PR

Any member of [core-adlr](https://github.com/orgs/teams/NVIDIA/core-adlr) and [`core-nemo`](https://github.com/orgs/teams/NVIDIA/core-nemo) will be able to merge your PR.

### CODE REVIEW

**Comment by FDecaYed** [comment]
@yanring @skyw

**Comment by skyw** [comment]
Could we add more tests with current broadcast path as reference? They should exact match after all communication is done.

File: megatron/core/optimizer/layer_wise_optimizer.py [hunk 1]
```diff
@@ -1,37 +1,39 @@
...
-    and grad.
+    2. Megatron DDP handle allreduce grad, note that each rank have full model and grad
     3. optimizer is already modified so only param belong to this DP rank is updated
-    3. grad_norm and zero counting will reduce metrics globally in step function
-    4. Do regular update with chained optimizers, optimizer is already modified so partial update
-    happens.
-    5. allgather updated params to every rank(currently through broadcast loop)
+    4. grad_norm and zero counting will reduce metrics globally in step function
+    5. Do regular update with chained optimizers, modified optimizer only update shard
+    6. allgather updated params to every rank(currently through broadcast loop)
```
Comment (skyw) [inline megatron/core/optimizer/layer_wise_optimizer.py]: Update comment. Not broadcast any more.

File: megatron/core/optimizer/layer_wise_optimizer.py [hunk 2]
```diff
@@ -53,40 +55,43 @@ def __init__(
...
         self.shard_params(optimizers)
-        # wrap optimizer after sharding to avoid unnecessary master weight creation
-        # TODO(deyuf): check if underlying optimizer.config need to fixed and if so can use
-        # that instead of passing
-        if init_state_fn_list is None:
-            init_state_fn_list = [None] * len(optimizers)
-        else:
+        if init_state_fn_list:
             assert len(init_state_fn_list) == len(optimizers), (
                 "init_state_fn_list must be the " "same length as optimizers if provided"
```
Comment (skyw) [inline megatron/core/optimizer/layer_wise_optimizer.py]: Quote is not necessary, it is probably left over from previous auto formatting.

File: megatron/core/optimizer/layer_wise_optimizer.py [hunk 3]
```diff
@@ -53,40 +55,43 @@ def __init__(
...
+            # unwrap FP32 optimizer, possibly from reusing get_megatron_optimizer for adam
+            for i, opt in enumerate(optimizers):
+                if isinstance(opt, Float16OptimizerWithFloat16Params):
+                    raise TypeError(
+                        'LayerWiseDistributedOptimizer received Float16 optimizer already.'
+                    )
+                # unwrap FP32 optimizer from reusing get_megatron_optimizer for adam
+                if isinstance(opt, FP32Optimizer):
+                    opt = opt.optimizer
+                optimizers[i] = Float16OptimizerWithFloat16Params(
```
Comment (skyw) [inline megatron/core/optimizer/layer_wise_optimizer.py:76]: Overriding list that is being looped is error prone and better to be avoided. One way (not the best) is to only loop index and access it by index. Or append to a new list.

File: tests/unit_tests/test_layer_wise_optimizer.py [hunk 4]
```diff
@@ -401,3 +402,39 @@ def test_parameter_updates_insufficient_parameters(self):
...
+
+        # Create reference model and optimizer using the same function
+        reference_model, reference_optimizer, _ = self.create_model_and_optimizer(
+            model_class=SimpleModel, copy_from=model
+        )
+
+        # Set same gradients on both models
+        for param, ref_param in zip(model.parameters(), reference_model.parameters()):
+            assert torch.equal(param.data, ref_param.data)
+            torch.testing.assert_close(param.data, ref_param.data, rtol=1e-5, atol=1e-5)
```
Comment (skyw) [inline tests/unit_tests/test_layer_wise_optimizer.py]: Does it pass with atol=0? It is safer if it can use relative only, i.e. none of the expected values are 0.

File: tests/unit_tests/test_layer_wise_optimizer.py [hunk 5]
```diff
@@ -401,3 +402,39 @@ def test_parameter_updates_insufficient_parameters(self):
...
+            grad_value = torch.randn_like(param)
```
Comment (skyw) [inline tests/unit_tests/test_layer_wise_optimizer.py:419]: It is easier to match with integer number.

File: tests/unit_tests/test_layer_wise_optimizer.py [hunk 6]
```diff
@@ -401,3 +402,39 @@ def test_parameter_updates_insufficient_parameters(self):
...
+
+        assert params_updated > 0, "At least some parameters should be updated"
+
+        # step() internal call allgather_params. replace reference object with bcast
+        reference_optimizer.allgather_params = reference_optimizer.broadcast_params
+        reference_optimizer.step()
+
+        # Verify updated values match reference optimizer
+        for param, ref_param in zip(model.parameters(), reference_model.parameters()):
+            torch.testing.assert_close(param.data, ref_param.data, rtol=1e-5, atol=1e-5)
```
Comment (skyw) [inline tests/unit_tests/test_layer_wise_optimizer.py]: Try to create exact match.
For example, creating both param and grad as random integers, use values that can be exactly represented as lr  , 0.5 or 2 for example. Then you have a great chance to match with both atol and rtol being 0.

Having both atol and rtol can be deceiving some time, e.g. if expected value is or the order of 1e-4, having a atol 1e-5 will hide a lot of errors.

File: tests/unit_tests/test_layer_wise_optimizer.py [hunk 7]
```diff
@@ -401,3 +402,39 @@ def test_parameter_updates_insufficient_parameters(self):
...
+        # Verify at least some parameters were updated
+        params_updated = 0
+        for param, ref_param in zip(model.parameters(), reference_model.parameters()):
+            if not torch.equal(param.data, ref_param.data):
+                params_updated += 1
+
+        assert params_updated > 0, "At least some parameters should be updated"
+
+        # step() internal call allgather_params. replace reference object with bcast
+        reference_optimizer.allgather_params = reference_optimizer.broadcast_params
```
Comment (skyw) [inline tests/unit_tests/test_layer_wise_optimizer.py:435]: How do we guarantee this takes effect?

I was thinking inherit a reference optimizer, and override allgather_params to call broadcast_params inside. If a check can guarantee the monkey patch has taken effect, that should also suffice.

Comment (FDecaYed) [reply tests/unit_tests/test_layer_wise_optimizer.py:435]: when adding this I had some debug logging and it was as expected(bcast called over allgather after patching). I think in python it is safe to assume patching the optimizer instance should take effect. if you still insist, I can add a flag to check or something

File: tests/unit_tests/test_layer_wise_optimizer.py [hunk 8]
```diff
@@ -401,3 +402,39 @@ def test_parameter_updates_insufficient_parameters(self):
...
+
+        # Create reference model and optimizer using the same function
+        reference_model, reference_optimizer, _ = self.create_model_and_optimizer(
+            model_class=SimpleModel, copy_from=model
+        )
+
+        # Set same gradients on both models
+        for param, ref_param in zip(model.parameters(), reference_model.parameters()):
+            assert torch.equal(param.data, ref_param.data)
+            torch.testing.assert_close(param.data, ref_param.data, rtol=0, atol=0)
```
Comment (skyw) [inline tests/unit_tests/test_layer_wise_optimizer.py:418]: Exact match? Nice :)

**Review by skyw — APPROVED** [review]
